tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs

```diff
@@ -0,0 +1,281 @@
+                    // c.f. https://learn.microsoft.com/sql/linux/sql-server-linux-docker-container-security?view=sql-server-ver16#storagepermissions
+
+                    // This is the minimal set to get the tests passing.
+                    const UnixFileMode MsSqlPermissions =
+                       UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute |
+                       UnixFileMode.OtherRead | UnixFileMode.OtherWrite | UnixFileMode.OtherExecute;
+
+                    File.SetUnixFileMode(bindMountPath, MsSqlPermissions);
+                    File.SetUnixFileMode($"{bindMountPath}/data", MsSqlPermissions);
+                    File.SetUnixFileMode($"{bindMountPath}/log", MsSqlPermissions);
+                    File.SetUnixFileMode($"{bindMountPath}/secrets", MsSqlPermissions);
+                }
+            }
+
+            using var app1 = builder1.Build();
+
+            await app1.StartAsync();
+
+            try
+            {
+                var hb1 = Host.CreateApplicationBuilder();
+
+                hb1.Configuration.AddInMemoryCollection(new Dictionary<string, string?>
+                {
+                    [$"ConnectionStrings:{masterdb1.Resource.Name}"] = await masterdb1.Resource.ConnectionStringExpression.GetValueAsync(default),
+                });
+
+                hb1.AddSqlServerClient(masterdb1.Resource.Name);
+
+                using var host1 = hb1.Build();
+
+                await host1.StartAsync();
+
+                await pipeline.ExecuteAsync(async token =>
+                {
+                    var conn = host1.Services.GetRequiredService<SqlConnection>();
+
+                    if (conn.State != System.Data.ConnectionState.Open)
+                    {
+                        await conn.OpenAsync(token);
+                    }
+
+                    var command = conn.CreateCommand();
+                    command.CommandText = """
+                        DROP TABLE IF EXISTS [Cars];
+                        CREATE TABLE [Cars] ([Brand] nvarchar(max) NOT NULL);
+                        INSERT INTO [Cars] ([Brand]) VALUES ('BatMobile');
+                        SELECT * FROM [Cars];
+                    """;
+
+                    var results = await command.ExecuteReaderAsync(token);
+
+                    Assert.True(results.HasRows);
+                }, cts.Token);
+
+                await app1.StopAsync();
+
+                await pipeline.ExecuteAsync(async token =>
+                {
+                    var conn = host1.Services.GetRequiredService<SqlConnection>();
+
+                    try
+                    {
+                        await conn.OpenAsync(token);
+                        Assert.Fail("Waiting for database to be down");
+                    }
+                    catch
+                    {
+                        // Failing means the database is correctly down
+                        return;
+                    }
+                }, cts.Token);
```

Comment:
I'm not sure I totally get what this chunk of code is trying to do.

Either the connection open will throw, or the next line `Assert.Fail` will throw. So either way an exception will occur, and that exception will be caught and eaten, and the code returns sucessfully.

Why do we have a "pipeline.ExecuteAsync" around this code? It will never retry.

Reply:
Right, the Fail is in the wrong place, I fixed the pattern somewhere else and missed this one.

We want the exception to be thrown if OpenAsync succeeds, so outside of the try/finally. The goal is to wait for the server to not respond to requests anymore.